satpy/modifiers/parallax.py

```diff
@@ -0,0 +1,537 @@
+        coordinates will also be in meter.
+    """
+    sat_xyz = np.hstack(lonlat2xyz(sat_lon, sat_lat)) * sat_alt
+    cth_xyz = np.stack(lonlat2xyz(lon, lat), axis=-1) * EARTH_RADIUS*1e3  # km → m
+    delta_xyz = cth_xyz - sat_xyz
+    sat_distance = np.sqrt((delta_xyz*delta_xyz).sum(axis=-1))
+    dist_shape = delta_xyz.shape[:-1] + (1,)  # force correct array broadcasting
+    return cth_xyz - delta_xyz*(parallax_distance/sat_distance).reshape(dist_shape)
+
+
+def _get_satellite_elevation(sat_lon, sat_lat, sat_alt, lon, lat):
+    """Get satellite elevation.
+
+    Get the satellite elevation from satellite lon/lat/alt for positions
+    lon/lat.
+    """
+    placeholder_date = datetime.datetime(2000, 1, 1)  # no impact on get_observer_look?
+    (_, elevation) = get_observer_look(
+            sat_lon, sat_lat, sat_alt/1e3,  # m → km (wanted by get_observer_look)
+            placeholder_date, lon, lat, 0)
+    return elevation
+
+
+def _calculate_parallax_distance(height, elevation):
+    """Calculate cloud to ground distance with parallax effect.
+
+    From (cloud top) height and satellite elevation, calculate the
+    cloud-to-ground distance taking the parallax effect into consideration.
+    """
+    if np.isscalar(elevation) and elevation == 0:
+        raise NotImplementedError(
+                "Parallax correction not implemented for "
+                "satellite elevation 0")
+    if np.isscalar(elevation) and elevation < 0:
+        raise ValueError(
+                "Satellite is below the horizon.  Cannot calculate parallax "
+                "correction.")
+    return height / np.sin(np.deg2rad(elevation))
+
+
+class ParallaxCorrection:
+    """Parallax correction calculations.
+
+    This class contains higher-level functionality to wrap the parallax
+    correction calculations in :func:`forward_parallax`.  The class is
+    initialised using a base area, which is the area for which a corrected
+    geolocation will be calculated.  The resulting object is a callable.
+    Calling the object with an array of (cloud top) heights returns a
+    :class:`~pyresample.geometry.SwathDefinition` describing the new ,
+    corrected geolocation.  The cloud top height should cover at least the
+    area for which the corrected geolocation will be calculated.
+
+    Note that the ``ctth`` dataset must contain satellite location
+    metadata, such as set in the ``orbital_parameters`` dataset attribute
+    that is set by many Satpy readers.  It is essential that the datasets to be
+    corrected are coming from the same platform as the provided cloud top
+    height.
+
+    A note on the algorithm and the implementation.  Parallax correction
+    is inherently an inverse problem.  The reported geolocation in
+    satellite data files is the true location plus the parallax error.
+    Therefore, this class first calculates the parallax error (using
+    :func:`forward_parallax`), which gives a shifted longitude and
```

Comment:
I currently use this code to compute the actual parallax displacement  

```
from pyproj import Geod
# Get parallax-corrected longitude and latitude 
(corr_lon, corr_lat) = forward_parallax(sat_lon, sat_lat, sat_alt, lon, lat, height)
# Get parallax displacement 
geod = Geod(ellps="WGS84")
_, _, parallax_dist = geod.inv(corr_lon, corr_lat, lon, lat)
```

Reply:
It gives the following parallax displacements for GOES-16 and GOES-17: 
 
![image](https://user-images.githubusercontent.com/19285200/181310450-eb2e7efe-b5b9-473a-aa8a-a98615585314.png)
![image](https://user-images.githubusercontent.com/19285200/181310477-637e6d17-e9f1-47df-83e9-5c939170de2f.png)

Maybe in the doc you could link also this [webpage](https://cimss.ssec.wisc.edu/goes/webapps/parallax/ ) which report the parallax displacement at 3 heights for GOES and Himawari

Reply:
That's wonderful.  I have added the utility function as you suggested, with the slight change that I've set the shape to be spherical, because a spherical Earth is also assumed in the pyorbital routines used to calculate the parallax error, so it's more consistent (and will not make much of a difference).

I think these figures may be nice to include in the documentation (one is enough).  What do you think?  We could link them directly, but it may be safer to upload them somewhere intended for permanent figure storage.  I've been uploading other images to figshare, but since you are the author here, maybe you might want to upload them?  Otherwise, I could recreate them.

Reply:
I can't currently run my codes to test it. But the code below should enable you to recreate the figure. 
Go ahead to upload the required figures on figshare ;) Thanks a lot again for this beautiful PR !!!

```
import satpy
import xarray as xr 
import cartopy.feature as cfeature
from cartopy import crs as ccrs
from satpy.modifiers.parallax import get_surface_parallax_displacement

area_def = satpy.resample.get_area_def("goes_east_abi_f_1km")
lon, lat = area_def.get_lonlats()

# Define satellite position    
sat_lat = 0
sat_lon = area_def.proj_dict['lon_0']
sat_alt = area_def.proj_dict['h']

# Define constant high clouds at 10 km 
height = np.full(lon.shape, 10_000)  

# Get parallax displacement 
parallax_dist = get_surface_parallax_displacement(sat_lon, sat_lat, sat_alt, lon, lat, height)

# Plot parallax displacement 
x_proj = area_def.projection_x_coords
y_proj = area_def.projection_y_coords
crs_ref = ccrs.PlateCarree()  
crs_proj = area_def.to_cartopy_crs()
da = xr.DataArray(parallax_dist,
                  dims=['y','x'],
                  coords={'x': x_proj, 
                          'y': y_proj},
                  )

figsize = (10,10)
fig, ax = plt.subplots(subplot_kw={'projection': crs_proj}, figsize=figsize, dpi=None)
p = da.plot.imshow(x='x', y='y', vmax=10000, ax=ax) 
ax.add_feature(cfeature.COASTLINE, edgecolor="black")
ax.set_title("GOES16 ABI Fixed Grid 1km - Parallax displacement with CTH=10km [m]")
plt.show()
```